[release-4.10] OCPBUGS-2013: Validate node subnet annotations against cluster networks

pperiyasamy opened the pull request:
This is manual cherry-pick of 4.11 commits a11e45dfc49537e506c661410b37b5285b0a893f 2d28afb74a330515f121e776f94de60b1000c1fe 711237766b25e0728006eacb8dc1478b1ce359bb 484c5a002c80c5b3dca1b67194dc5271ba46acce 3eb7a5e7bc8e2dc7b8fef4762dd96b1cd20c6fd4.

Signed-off-by: Periyasamy Palanisamy <pepalani@redhat.com>

pperiyasamy commented:
/jira refresh

pperiyasamy commented:
/jira refresh

pperiyasamy commented:
/retest

pperiyasamy commented:
/retest

pperiyasamy commented:
/retest

pperiyasamy commented:
/jira refresh

pperiyasamy commented:
/jira refresh

pperiyasamy commented:
/assign @dcbw

pperiyasamy commented:
/retest

pperiyasamy commented:
/jira refresh

dcbw commented:
/hold

Windows team reported regression in the 4.11 backport which I think this PR will also be subject to. @JacobTanenbaum is figuring out the fix.

pperiyasamy commented:
/retest-required

pperiyasamy commented:
/retest-required

dcbw commented:
I think we can unhold this now that it's fixed...

/hold cancel

dcbw commented:
/approve

@JacobTanenbaum can you LGTM if it looks OK to you now?

dcbw commented:
/test unit

pperiyasamy commented:
/test unit

JacobTanenbaum commented:
/lgtm

jcaamano commented on go-controller/pkg/ovn/ovn.go:
```diff
@@ -1362,8 +1362,7 @@ func (oc *Controller) WatchNodes() {
 
-			nodeSubnets, _ := util.ParseNodeHostSubnetAnnotation(node)
-			oc.deleteNode(node.Name, nodeSubnets)
+			oc.deleteNode(node.Name)
```
Should we Klog something here on error?

jcaamano commented on go-controller/pkg/ovn/master.go:
```diff
@@ -1236,15 +1097,7 @@ func (oc *Controller) deleteNodeLogicalNetwork(nodeName string) error {
-	}
-	// update metrics
-	metrics.RecordSubnetUsage(oc.v4HostSubnetsUsed, oc.v6HostSubnetsUsed)
+	oc.masterSubnetAllocator.ReleaseNodeSubnets(nodeName, hostSubnets...)
```
OK, I see that a later commit changes this to not do anything on error. All fine (although having a klog warning would have been reasonable as well).

pperiyasamy replied:
yes, done.

jcaamano reviewed (commented):
Two things in addition to the comments:

* For commit `[ovn: ensure existing node subnet annotations are valid when a node is added](https://github.com/openshift/ovn-kubernetes/pull/1332/commits/ed408a58c6c76feff6bf7e851170ff916c8e773e)` I am missing some changes in `master_test.go`. Is that intentional?
* Commit https://github.com/openshift/ovn-kubernetes/pull/1332/commits/79141ba7f4ac2ecdf60c931a0f68edfea8684845 has also nothing to do with the cherry-picked commit, its completely different and unrelated. So I am missing something here as well.

pperiyasamy commented:
> Two things in addition to the comments:
> 
> * For commit `[ovn: ensure existing node subnet annotations are valid when a node is added](https://github.com/openshift/ovn-kubernetes/pull/1332/commits/ed408a58c6c76feff6bf7e851170ff916c8e773e)` I am missing some changes in `master_test.go`. Is that intentional?
> * Commit [79141ba](https://github.com/openshift/ovn-kubernetes/commit/79141ba7f4ac2ecdf60c931a0f68edfea8684845) has also nothing to do with the cherry-picked commit, its completely different and unrelated. So I am missing something here as well.

Yes @jcaamano The commit for [79141ba](https://github.com/openshift/ovn-kubernetes/commit/79141ba7f4ac2ecdf60c931a0f68edfea8684845) doesn't have changes to master.go because on 4.10 it doesn't have `addUpdateNodeEvent` implementation, hence ignored `recordNodeErrorEvent` changes and also used it as placeholder to get required fixes in master_test.go (as per refactoring done in 4.11).  klog and other stuffs just came up while cherry-picking actual 4.11 commits.

pperiyasamy commented on go-controller/pkg/ovn/master_test.go:
```diff
@@ -1217,289 +1216,68 @@ var _ = ginkgo.Describe("Gateway Init Operations", func() {
 }
 
-func TestController_syncNodesRetriable(t *testing.T) {
+func TestController_syncNodes(t *testing.T) {
```
These changes came from 4.11 at cherry-pick, removed it now.

pperiyasamy commented:
> > Two things in addition to the comments:
> > 
> > * For commit `[ovn: ensure existing node subnet annotations are valid when a node is added](https://github.com/openshift/ovn-kubernetes/pull/1332/commits/ed408a58c6c76feff6bf7e851170ff916c8e773e)` I am missing some changes in `master_test.go`. Is that intentional?
> > * Commit [79141ba](https://github.com/openshift/ovn-kubernetes/commit/79141ba7f4ac2ecdf60c931a0f68edfea8684845) has also nothing to do with the cherry-picked commit, its completely different and unrelated. So I am missing something here as well.
> 
> Yes @jcaamano The commit for [79141ba](https://github.com/openshift/ovn-kubernetes/commit/79141ba7f4ac2ecdf60c931a0f68edfea8684845) doesn't have changes to master.go because on 4.10 it doesn't have `addUpdateNodeEvent` implementation, hence ignored `recordNodeErrorEvent` changes and also used it as placeholder to get required fixes in master_test.go (as per refactoring done in 4.11). klog and other stuffs just came up while cherry-picking actual 4.11 commits.



> > Two things in addition to the comments:
> > 
> > * For commit `[ovn: ensure existing node subnet annotations are valid when a node is added](https://github.com/openshift/ovn-kubernetes/pull/1332/commits/ed408a58c6c76feff6bf7e851170ff916c8e773e)` I am missing some changes in `master_test.go`. Is that intentional?
> > * Commit [79141ba](https://github.com/openshift/ovn-kubernetes/commit/79141ba7f4ac2ecdf60c931a0f68edfea8684845) has also nothing to do with the cherry-picked commit, its completely different and unrelated. So I am missing something here as well.
> 
> Yes @jcaamano The commit for [79141ba](https://github.com/openshift/ovn-kubernetes/commit/79141ba7f4ac2ecdf60c931a0f68edfea8684845) doesn't have changes to master.go because on 4.10 it doesn't have `addUpdateNodeEvent` implementation, hence ignored `recordNodeErrorEvent` changes and also used it as placeholder to get required fixes in master_test.go (as per refactoring done in 4.11). klog and other stuffs just came up while cherry-picking actual 4.11 commits.

@jcaamano For 1, rebased the commit https://github.com/openshift/ovn-kubernetes/pull/1332/commits/19159d9f9b63652aa0080cb58bb1f2432326ee74 to have those missing test changes. For 2, Dropped out cherry-pick of https://github.com/openshift/ovn-kubernetes/commit/3eb7a5e7bc8e2dc7b8fef4762dd96b1cd20c6fd4 commit to this PR as it's not needed for 4.10. Hope it makes sense now.

pperiyasamy commented:
/retest-required

jcaamano commented:
/lgtm
/approve
/label backport-risk-assessed

anuragthehatter commented:
/label cherry-pick-approved

pperiyasamy commented:
/retest-required

pperiyasamy commented:
/retest-required

pperiyasamy commented:
/retest-required

pperiyasamy commented:
/retest-required

pperiyasamy commented:
/retest-required
